[CBRD-25542] Problem with partition pruning when using multi-column indexes with IN and range predicates

Hamkua opened the pull request:
http://jira.cubrid.org/browse/CBRD-25542

Partition pruning does not work correctly in the following cases:

- When the WHERE clause includes the following conditions:

    - An IN condition on the partition key column A

    - A range condition (e.g., > or <) on column B, which forms the same index as the partition key column A, without an equality condition on column B

- Additionally, if the partitioned table has an index consisting of columns A, B, and others, and this index is used to scan the partitioned table.

The issue is caused by the use of a multi-column index.

With the addition of the MVCC feature, the `where_range` now contains a `PRED_EXPR` that compares a single column of the index. This has been modified to utilize `where_range` for partition pruning.

Hamkua committed:
[CBRD-25542] modify partition pruning to use WHERE_RANGE PRED_EXPR instead of indexptr

Hamkua committed:
Merge branch 'CUBRID:develop' into CBRD-25542

shparkcubrid commented on src/query/partition.c:
```diff
@@ -2887,45 +2889,41 @@ partition_prune_spec (THREAD_ENTRY * thread_p, val_descr * vd, access_spec_node
+    }
 
-      if (pinfo.partition_pred->func_regu->type != TYPE_ATTR_ID)
+  if (status == MATCH_NOT_FOUND)
```
찾지 못한 경우를 에러처리 해야되나요?

Hamkua replied:
아래의 함수에서 문제가 발생하는 경우 pruning_context에 error_code를 설정함과 동시에, MATCH_NOT_FOUND를 리턴하기 때문입니다.

`partition_match_pred_expr`

- `partition_prune`

  - `partition_get_value_from_regu_var` : 상수 값 또는 pseudo constants 가 아닌 regu_variable로부터 db_value를 추출하려고 시도하는 경우
    
  - `partition_prune_db_val`
       - `partition_prune_hash`
       - `partition_prune_range` 
       - `partition_prune_list` :  pruning set의 size보다 크거나, 0보다 작은 인덱스를 사용하는 경우

Hamkua commented on src/query/partition.c:
```diff
@@ -2887,45 +2889,41 @@ partition_prune_spec (THREAD_ENTRY * thread_p, val_descr * vd, access_spec_node
-	  partition_clear_pruning_context (&pinfo);
-	  return ER_FAILED;
-	}
+  if (pinfo.spec->where_range != NULL)
```
PRUNING_BITSET을 덮어쓰기 때문에, 이전에 프루닝된 정보가 제거되는 문제가 있습니다.
기존 큐브리드에서도 동일하게 발생하는 문제입니다.

예시 쿼리: 
```
create table t1(c1 int, index i1(c1)) partition by range(c1) (
    partition p1 values less than (10),
    partition p2 values less than (20),
    partition p3 values less than (30),
    partition p4 values less than (40)
);
insert into t1 values (0), (9), (10), (19), (20), (29), (30), (39);

/* 
where_key : c1 < mod(rand(), 10) + 20
where_range : c1 > 10
*/
select /*+ recompile */ * from t1 where c1 > 10 and c1 < mod(rand(), 10) + 20;
```
where_key에서 p1, p2, p3 파티션이 선택됩니다. ( 20<= mod(rand(), 10) + 20 < 30 )
where_range에서 p2, p3, p4 파티션이 선택됩니다.

Hamkua committed:
[CBRD-25542] modify the pruning set to compute the intersection instead of overwriting it

Hamkua committed:
Merge branch 'CUBRID:develop' into CBRD-25542

Hamkua committed:
Merge branch 'CUBRID:develop' into CBRD-25542

hgryoo committed:
[CBRD-25106] Slip: updating target_method column during ALTER PROCEDURE .. COMPILE (#5728)

http://jira.cubrid.org/browse/CBRD-25106

Fix target_method column is not updated

Hamkua committed:
Merge remote-tracking branch 'upstream/feature/partition-table' into CBRD-25542

Hamkua committed:
[CBRD-25542] Set MATCH_OK status when partition pruning is applied

Hamkua committed:
[CBRD-25542] move pruning bitset intersection logic to partition_prune_spec

Hamkua committed:
[CBRD-25542] add comments to partition_prune_spec function

H2SU committed:
[CBRD-25760] Fix grep command in check.yml to prevent false matches (#5730)

http://jira.cubrid.org/browse/CBRD-25760

YeunjunLee committed:
[CBRD-24741] Provides a way to revive the cub_server process in case of abnormal shutdown (#5740)

http://jira.cubrid.org/browse/CBRD-24741

hyunikn committed:
[CBRD-25750] coerce SP parameter default values to VARCHAR(255) before saving it (#5736)

coerce SP parameter default values to VARCHAR(255) before saving it to _db_stored_procedure_args.default_value

Hamkua committed:
Merge branch 'CUBRID:develop' into CBRD-25542

Hamkua committed:
[CBRD-25542] revert changes except partition.c

Hamkua committed:
[CBRD-25542] revert changes except partition.c

shparkcubrid commented on src/query/partition.c:
```diff
@@ -2888,54 +2890,78 @@ partition_prune_spec (THREAD_ENTRY * thread_p, val_descr * vd, access_spec_node
+	      goto error_exit;
 	    }
 	}
-      error = partition_prune_index_scan (&pinfo);
```
partition_prune_heap_scan () 와 partition_prune_index_scan ()를 '#if defined (ENABLE_UNUSED_FUNCTION)' 해주세요.

Hamkua replied:
네, 추가했습니다.

Hamkua committed:
[CBRD-25542] mark partition_prune_heap_scan and partition_prune_index_scan as unused functions

Hamkua committed:
[CBRD-25542] mark unused function declarations with ENABLE_UNUSED_FUNCTION

youngjinj commented on src/query/partition.c:
```diff
@@ -2702,6 +2702,7 @@ partition_get_position_in_key (PRUNING_CONTEXT * pinfo, BTID * btid)
   return NO_ERROR;
 }
 
+#if defined (ENABLE_UNUSED_FUNCTION)
```
partition_prune_heap_scan, partition_prune_index_scan 함수 선언에도 ENABLE_UNUSED_FUNCTION를 정의해주세요.

```cpp
#if defined (ENABLE_UNUSED_FUNCTION)
static int partition_prune_heap_scan (PRUNING_CONTEXT * pinfo);
static int partition_prune_index_scan (PRUNING_CONTEXT * pinfo);
#endif /* ENABLE_UNUSED_FUNCTION */
```

Hamkua replied:
네, 함수 선언부에도 추가했습니다.
